pkg/cryptoauthlib/contrib/atca.c

```diff
@@ -0,0 +1,204 @@
+/*
+ * Copyright (C) 2019 2019 HAW Hamburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @ingroup     pkg_cryptoauthlib cryptoauthlib security crypto
+ * @{
+ *
+ * @file
+ * @brief       HAL implementation for the library Microchip CryptoAuth devices
+ *
+ * @author      Lena Boeckmann <lena.boeckmann@haw-hamburg.de>
+ *
+ * @}
+ */
+
+#include <stdint.h>
+#include "xtimer.h"
+#include "periph/i2c.h"
+#include "periph/gpio.h"
+
+#include "atca.h"
+#include "atca_params.h"
+
+
+/* Timer functions */
+void atca_delay_us(uint32_t delay)
+{
+    xtimer_usleep(delay);
+}
+
+void atca_delay_10us(uint32_t delay)
+{
+    xtimer_usleep(delay * 10);
+}
+
+void atca_delay_ms(uint32_t delay)
+{
+    xtimer_usleep(delay * 1000);
+}
+
+/* Hal I2C implementation */
+ATCA_STATUS hal_i2c_init(void *hal, ATCAIfaceCfg *cfg)
+{   
+    if(cfg->iface_type != ATCA_I2C_IFACE)
+    {
+        return ATCA_BAD_PARAM;
+    }
+    
+    ((ATCAHAL_t*)hal)->hal_data = cfg;
+    
+    atcab_wakeup();
+
+    return ATCA_SUCCESS;
+}
+
+ATCA_STATUS hal_i2c_post_init(ATCAIface iface)
+{
+    return ATCA_SUCCESS;
+}
+
+ATCA_STATUS hal_i2c_send(ATCAIface iface, uint8_t *txdata, int txlength)
+{
+    ATCAIfaceCfg *cfg = atgetifacecfg(iface);
+    int ret = -1; 
+    int txlength_updated = txlength + 1;
+    /* First byte in command packages is reserved for hal use as needed
+    We use it for the word address */
+    txdata[0] = ATCA_DATA_ADR;
+
+    /* slave address needs to be shifted by 1 to ignore lsb (rw bit) */
```

Comment:
It doesn't happen automatically. I need to shift the slave address by one when writing to and reading from the bus. Otherwise the address will be incorrect. Of course I can just do it without commenting on it.